### Import necessary libraries to demonstrate API calls

In [1]:
import json
import requests
import os
import uuid

### **Note:** if unexpected/unhandled server-side error occurs, the following will be returned:

{'error': 'An unexpected error occurred'}, 500

# Receipt API Endpoints (Local Only, Work-In-Progress)

### receipts_parsing (POST): Parses a receipt image and returns parsed receipt

**Parameters:**

**Required:**
files = {'image':{binary image}}


**Response:**

If success:
{
    'receipt_date': {date on receipt},
    'total': {total},
    'store': {store name},
    'location': {store address},
    'purchases': [{'name':{name of item},'price':{unit price},'amount':{amount/quantity}},{'name':{name of item},'price':{unit price},'amount':{amount/quantity}}]
}

If fail (see error.py for details): MissingReceiptImage(), InvalidReceiptImage()

In [2]:
### Login locally to test local endpoints (endpoints not on PythonAnywhere server)
url = 'http://127.0.0.1:5000/api/login'
returning_user = {'user_id':'Dialga22',
          'password':'Dialga22'}
response = requests.post(url, json=returning_user)
print('Response code: ' + str(response.status_code))
print(json.dumps(response.json(), indent=4))
print(response.cookies)
session_cookie = response.cookies

Response code: 200
{
    "message": "Dialga22 logged in successfully."
}
<RequestsCookieJar[<Cookie session=.eJwlzrsNAjEMANBdUlPYiT_x1eyBHMcBJKqDqxC7g8Sb4L3LZe35vJXttR95Kpf7LFuZadWTQmhpdOXW1mJJEaeWXa2N5Rk-MsQwJ4OGzZjsNmLBqiaRyE2NoNIgZIsOaNZZTVFGFxJuHad0tCorwRs4DgJgUrPyixzP3P-b890fV6-1fL5D9TD1.Zq-hDg.T7D_sNS2PWIXaDo4gZ-1Lb1HzVU for 127.0.0.1/>]>


In [3]:
url = 'http://127.0.0.1:5000/api/receipts_parsing'
image_path = os.path.join(os.getcwd(), "sample_receipts/QFC_1.jpg")

# Generate a random filename (helps avoid collisions when saving/deleting files on server in the backend)
random_filename = f"{uuid.uuid4()}.jpg"

# Open the image file in binary mode
with open(image_path, 'rb') as image_file:
    # Create a dictionary to store the file data
    files = {
        'receipt_image': (
            random_filename,  # Name of the file
            image_file,   # File object
            'image/png'   # MIME type
        )
    }
    response = requests.post(url, files=files, cookies = session_cookie)


print('Response code: ' + str(response.status_code))
print(json.dumps(response.json(), indent=4))


Response code: 201
{
    "amazon": {
        "cost": 0.01,
        "extracted_data": [
            {
                "bank": {
                    "account_number": null,
                    "bic": null,
                    "bsb": null,
                    "iban": null,
                    "routing_number": null,
                    "sort_code": null,
                    "swift": null
                },
                "customer_information": {
                    "abn_number": null,
                    "billing_address": null,
                    "business_number": null,
                    "city": null,
                    "coc_number": null,
                    "country": null,
                    "customer_number": null,
                    "email": null,
                    "fiscal_number": null,
                    "gst_number": null,
                    "house_number": null,
                    "id_reference": null,
                    "mailling_address": null,
                 

# Receipt API Endpoints (Deployed on PythonAnywhere)

### register (Post): User registration
**Parameters:** 

{**"user_id"**:{user_id}, **"password"**:{password}, **"email"**:{email}, **"date"**:{date on users device}}

**Response:**

If success: {"message": "{current user} logged in successfully."}, 200

If fail (see error.py for details): MissingUserIDError(), MissingPasswordError(), MissingEmailError(), MissingUserDate(), InvalidDateFormat(), UserAlreadyExistsError(), EmailAlreadyExistsError() 


In [42]:
# url = 'http://127.0.0.1:5000/api/register'
url = 'https://receiptplus.pythonanywhere.com/api/register'
new_user = {'user_id':'Dialga226',
          'password':'Dialga226',
          'email': "Dialga226@gmail.com",
          'date':'2024-07-06'}
response = requests.post(url, json=new_user)
print('Response code: ' + str(response.status_code))
print(json.dumps(response.json(), indent=4))
session_cookie = response.cookies

Response code: 400
{
    "error": "User already exists"
}


### login (Post): User Login
**Parameters:**

{**"user_id"**:{user_id}, **"password"**:{password}} *or* {**"user_id"**:{email}, **"password"**:{password}}

**Response:**

If success: {"message": "{current user} logged in successfully."}, 200

If fail (see error.py for details): MissingUserIDError(), MissingPasswordError(), UserNotFound(), InvalidPassword()


In [43]:
url = 'https://receiptplus.pythonanywhere.com/api/login'
# url = 'http://127.0.0.1:5000/api/login'
returning_user = {'user_id':'Dialga226',
          'password':'Dialga226'}
response = requests.post(url, json=returning_user)
print('Response code: ' + str(response.status_code))
print(json.dumps(response.json(), indent=4))
print(response.cookies)
session_cookie = response.cookies

Response code: 200
{
    "message": "Dialga226 logged in successfully."
}
<RequestsCookieJar[<Cookie session=.eJwlzjsOwjAMANC7ZGZI4l_SmXtUdmJDJaaWToi7U4kLPL1PWmP345mW9376La3bTEvC3inLLFGEkbpTKx6VSXi0NqPmkMkeBhPLxDa0o6p5RLOWsUlmIA6e4iARTp3ECigYuqGAidAgZwAZOi6tGM6hphmiFwhLV-Q8fP9v7pu-Hlorp-8PTsgymQ.Zq553g.9bYNfJOS-S_QH6EwjqDtQ8CTZyc for receiptplus.pythonanywhere.com/>]>


### receipts (POST): User adds a receipt

**Parameters:**

**Required:**
{'receipt_date':{date on receipt}, 'total':{receipt total}}

**Optional:**

{'category':{category of receipt}, 'store':{store name}, 'location':{store address}, 

'purchases': [{'name':{item 1 name}, 'price':{item 1 price}, 'amount':{item 1 quantity or weight}}, {'name':{item 2 name}, 'price':{item 2 price}, 'amount':{item 2 quantity or weight}}...]}

**Response:**

If success: {'message': 'Receipt uploaded successfully.'}), 201

If fail (see error.py for details): MissingReceiptDate(), MissingReceiptTotal()

In [44]:
url = 'https://receiptplus.pythonanywhere.com/api/receipts'
# url = 'http://127.0.0.1:5000/api/receipts'
receipt = {
    'receipt_date': "2024-07-07",
    'category': 'Groceries',
    'total': '23.33',
    'store': 'QFC',
    'location': '4500 Wallingford Ave N, Seattle, WA 98103',
    'purchases': [{'name':'banana'},{'name':'leaves','price':4.99,'amount':6},{'name':'pineapple','price':5.99,'amount':6}]
}

# seesion_cookie = ''
response = requests.post(url, json=receipt, cookies = session_cookie)
print('Response code: ' + str(response.status_code))
print(json.dumps(response.json(), indent=4))

Response code: 201
{
    "message": "Receipt uploaded successfully."
}


In [45]:
url = 'https://receiptplus.pythonanywhere.com/api/receipts'
# url = 'http://127.0.0.1:5000/api/receipts'
receipt = {
    'receipt_date': "2024-07-08",
    'category': 'Gas',
    'total': '63.33',
    'store': 'Chevron',
    'location': '427 12th Ave, Seattle, WA 98122',
    'purchases': [{'name':'gas','price':3.99,'amount':18}]
}

response = requests.post(url, json=receipt, cookies = session_cookie)
print('Response code: ' + str(response.status_code))
print(json.dumps(response.json(), indent=4))

Response code: 201
{
    "message": "Receipt uploaded successfully."
}


### month_cat_exp (GET): Gets user spending per category for current month

**Parameters:** 

{**'date'**:{date on users device}}

**Response:** 

If success: {'{category 1}': {spending 1}, '{category 2}':{spending 2},...}

If fail (see error.py for details): MissingUserDate(), InvalidDateFormat()

In [46]:
# url = 'http://127.0.0.1:5000/api/month_cat_exp'
url = 'https://receiptplus.pythonanywhere.com/api/month_cat_exp'
response = requests.get(url, json= {'date': '2024-07-07'}, cookies = session_cookie)
print('Response code: ' + str(response.status_code))
print(json.dumps(response.json(), indent=4))

Response code: 200
{
    "Gas": 126.66,
    "Groceries": 46.66
}


### month_exp (GET): Gets the days the user spent money on and the amount spent on those days for the current month

**Parameters:**

{**'date'**:{date on users device}}

**Response:**

If success:{'{day n of month}': {spending on day n}, '{day m of month}': {spending on day m},...}

If fail (see error.py for details): MissingUserDate(), InvalidDateFormat()

In [47]:
url = 'https://receiptplus.pythonanywhere.com/api/month_exp'
# url = 'http://127.0.0.1:5000/api/month_exp'
response = requests.get(url, json= {'date': '2024-07-07'}, cookies = session_cookie)
print('Response code: ' + str(response.status_code))
print(json.dumps(response.json(), indent=4))

Response code: 200
{
    "7": 46.66,
    "8": 126.66
}


## User Info

In [48]:
# session_cookie = ".eJwlzjsOwjAMANC7ZGaIndipexnkXwQSQ9XSCXF3KjG_5X3Kfe55PMr63s-8lfszylogqVZBp8GcyMJCmL2Z9lzCghQVh6bIZRFTzYCQclRhS1NRiUnhrYtDU1eoIxcOtVTgBRqFzjalpnvHhYJHzN7YrZNU71CuyHnk_t_otr0SsJXvD1IGMqY.Zq1Zyg.yS_mJxHOVBlIajDSmS6dFJD4hiw"
url = 'https://receiptplus.pythonanywhere.com/api/user_info'
# url = 'http://127.0.0.1:5000/api/user_info'
# response = requests.get(url, cookies = session_cookie)
response = requests.get(url, cookies = session_cookie)
print(json.dumps(response.json(), indent=4))

{
    "date_joined": "2024-07-06",
    "email": "Dialga226@gmail.com",
    "user_id": "Dialga226"
}


### Logout (Post): User Logout
**Parameters:** Session ID

**Response Format:**


In [49]:
url = 'https://receiptplus.pythonanywhere.com/api/logout'
response = requests.post(url, cookies = session_cookie)
print(response.cookies)
cookies = response.cookies

<RequestsCookieJar[]>


## Figma for Android/iOS App: [Receipt+ Figma](https://www.figma.com/design/1hubQrhZHVsg3J9vlG2Kkg/Receipt%2B?node-id=0%3A1&t=ZNRuRUXyLTjp1hEQ-1)

### Receipt (GET): Gets all user receipts (Not completed, work in Progress...)

**Parameters:** {'Start':'MM-YYYY', 'End':'MM-YYYY'}

**Response:** 

List of jsonified receipt objects with additional receipt id in the given time range